packages/apollo-server-core/src/requestPipeline.ts

```diff
@@ -221,33 +245,66 @@ export async function processGraphQLRequest<TContext>(
     );
 
     let response: GraphQLResponse;
+    let result: ExecutionResult | DeferredExecutionResult;
+    let patches: AsyncIterable<ExecutionPatchResult> | undefined;
+    let isDeferred = false;
 
     try {
-      response = (await execute(
+      result = (await execute(
         document,
         request.operationName,
         request.variables,
-      )) as GraphQLResponse;
-      executionDidEnd();
-    } catch (executionError) {
-      executionDidEnd(executionError);
-      return sendErrorResponse(executionError);
-    }
+      ));
 
-    const formattedExtensions = extensionStack.format();
-    if (Object.keys(formattedExtensions).length > 0) {
-      response.extensions = formattedExtensions;
-    }
+      isDeferred = isDeferredExecutionResult(result);
 
-    if (config.formatResponse) {
-      response = config.formatResponse(response, {
+      if (isDeferred) {
+        response = ((result as DeferredExecutionResult).initialResult) as GraphQLResponse;
+        patches = (result as DeferredExecutionResult).deferredPatches;
+      } else {
+        response = result as GraphQLResponse;
+      }
+
+      const formattedExtensions = extensionStack.format();
+      if (Object.keys(formattedExtensions).length > 0) {
+        response.extensions = formattedExtensions;
+      }
+
+      // `formatResponse` format fallback for TS2722: Cannot invoke an object which is possibly 'undefined'.
+      const formatResponse = config.formatResponse || ((x: GraphQLResponse):GraphQLResponse => x);
+
+      response = formatResponse(response, {
         context: requestContext.context,
       });
+
+      let output: GraphQLResponse | DeferredGraphQLResponse;
+
+      if (isDeferred) {
+        executionDidEnd();
+        output = {
+          initialResponse: response,
+          deferredPatches: patches!,
+          requestDidEnd,
+        };
+
+      } else {
+
+        executionDidEnd();
+        output = response;
```

Comment:
Did you consider always returning a`DeferredGraphQLResponse` so that you do not have the `isDeferred` statements everywhere? Or would that create too much performance overhead because it turns everything into an Iterable? Maybe that could be avoided by calling the `requestDidEnd` right away when not deferred as you do on line 306.

I think that would clean up a lot of (close to) duplicate code behind all the `isDeferred` checks.